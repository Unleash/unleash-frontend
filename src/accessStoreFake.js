import { Map as $MAp } from 'immutable';

export const createFakeStore = (permissions) => {
    return {
        getState: () => ({
            user: 
                new $MAp({
                    permissions
                })
        }),
    }
}