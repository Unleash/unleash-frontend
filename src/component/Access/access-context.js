import React from 'react';

const AccessContext = React.createContext()

const AccessProvider = ({store, children}) => {
  const hasAccess = (permission, project) => {
    const permissions = store.getState().user.get('permissions');

    return permissions.some(p => {
      if(p.permission === 'ADMIN') {
        return true
      }
      if(p.permission === permission && p.project === project) {
        return true;
      }
      return false;
    })
    
  };

  const context = { hasAccess };

  return <AccessContext.Provider value={context}>{children}</AccessContext.Provider>
}

export {Â AccessProvider };
export default AccessContext;
