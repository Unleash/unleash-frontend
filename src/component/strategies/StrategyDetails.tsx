import { useContext, useEffect } from 'react';
import PropTypes from 'prop-types';
import { Grid, Typography } from '@material-ui/core';
import ShowStrategy from './show-strategy-component';
import EditStrategy from './CreateStrategy';
import { UPDATE_STRATEGY } from '../providers/AccessProvider/permissions';
import ConditionallyRender from '../common/ConditionallyRender/ConditionallyRender';
import TabNav from '../common/TabNav/TabNav';
import PageContent from '../common/PageContent/PageContent';
import AccessContext from '../../contexts/AccessContext';
import useStrategies from '../../hooks/api/getters/useStrategies/useStrategies';
import { useParams } from 'react-router-dom';

const StrategyDetails = props => {
    const { hasAccess } = useContext(AccessContext);
    const { strategies, refetchStrategies } = useStrategies();
    const { strategyName } = useParams<{ strategyName: string }>();
    const strategy = strategies.find(n => n.name === strategyName);
    
    useEffect(() => {
        if (!strategy) {
            refetchStrategies();
        }
        if (!props.applications || props.applications.length === 0) {
            props.fetchApplications();
        }
        if (!props.toggles || props.toggles.length === 0) {
            props.fetchFeatureToggles();
        }
        //eslint-disable-next-line
    }, []);

    const tabData = [
        {
            label: 'Details',
            component: (
                <ShowStrategy
                    strategy={strategy}
                    toggles={props.toggles}
                    applications={props.applications}
                />
            ),
        },
        {
            label: 'Edit',
            component: (
                <EditStrategy
                    strategy={strategy}
                    history={props.history}
                    editMode
                />
            ),
        },
    ];

    if (!strategy) return null;
    return (
        <PageContent headerContent={strategy.name}>
            <Grid container>
                <Grid item xs={12} sm={12}>
                    <Typography variant="subtitle1">
                        {strategy.description}
                    </Typography>
                    <ConditionallyRender
                        condition={
                            strategy.editable && hasAccess(UPDATE_STRATEGY)
                        }
                        show={
                            <div>
                                <TabNav tabData={tabData} />
                            </div>
                        }
                        elseShow={
                            <section>
                                <div className="content">
                                    <ShowStrategy
                                        strategy={strategy}
                                        toggles={props.toggles}
                                        applications={props.applications}
                                    />
                                </div>
                            </section>
                        }
                    />
                </Grid>
            </Grid>
        </PageContent>
    );
};

StrategyDetails.propTypes = {
    toggles: PropTypes.array,
    applications: PropTypes.array,
    activeTab: PropTypes.string.isRequired,
    fetchStrategies: PropTypes.func.isRequired,
    fetchApplications: PropTypes.func.isRequired,
    fetchFeatureToggles: PropTypes.func.isRequired,
    history: PropTypes.object.isRequired,
};

export default StrategyDetails;
