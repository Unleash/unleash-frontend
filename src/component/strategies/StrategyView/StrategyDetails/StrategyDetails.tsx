import {
    Grid,
    List,
    ListItem,
    ListItemAvatar,
    ListItemText,
    Tooltip,
} from '@material-ui/core';
import { Add, RadioButtonChecked } from '@material-ui/icons';
import { AppsLinkList } from 'component/common';
import ConditionallyRender from 'component/common/ConditionallyRender';
import styles from '../../strategies.module.scss';
import { TogglesLinkList } from 'component/strategies/TogglesLinkList/TogglesLinkList';
import { IParameter, IStrategy } from 'interfaces/strategy';
import { IApplication } from 'interfaces/application';
import { IFeatureToggle } from 'interfaces/featureToggle';

interface IStrategyDetailsProps {
    strategy: IStrategy;
    applications: IApplication[];
    toggles: IFeatureToggle[];
}

export const StrategyDetails = ({
    strategy,
    applications,
    toggles,
}: IStrategyDetailsProps) => {
    const { parameters = [] } = strategy;
    const renderParameters = (params: IParameter[]) => {
        if (params.length > 0) {
            return params.map(({ name, type, description, required }, i) => (
                <ListItem key={`${name}-${i}`}>
                    <ConditionallyRender
                        condition={required}
                        show={
                            <ListItemAvatar>
                                <Tooltip title="Required parameter">
                                    <Add aria-hidden={false} />
                                </Tooltip>
                            </ListItemAvatar>
                        }
                        elseShow={
                            <ListItemAvatar>
                                <Tooltip title="Optional parameter">
                                    <RadioButtonChecked aria-hidden={false} />
                                </Tooltip>
                            </ListItemAvatar>
                        }
                    />
                    <ListItemText
                        primary={
                            <div>
                                {name} <small>({type})</small>
                            </div>
                        }
                        secondary={description}
                    />
                </ListItem>
            ));
        } else {
            return <ListItem>No params</ListItem>;
        }
    };

    return (
        <div className={styles.listcontainer}>
            <Grid container>
                <ConditionallyRender
                    condition={strategy.deprecated}
                    show={
                        <Grid item>
                            <h5 style={{ color: '#ff0000' }}>Deprecated</h5>
                        </Grid>
                    }
                />
                <Grid item sm={12} md={12}>
                    <h6>Parameters</h6>
                    <hr />
                    <List>{renderParameters(parameters)}</List>
                </Grid>

                <Grid item sm={12} md={6}>
                    <h6>Applications using this strategy</h6>
                    <hr />
                    <AppsLinkList apps={applications} />
                </Grid>

                <Grid item sm={12} md={6}>
                    <h6>Toggles using this strategy</h6>
                    <hr />
                    <TogglesLinkList toggles={toggles} />
                </Grid>
            </Grid>
        </div>
    );
};
