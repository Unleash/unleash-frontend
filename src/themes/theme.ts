import { createTheme } from '@mui/material/styles';
import { colors } from './colors';

export default createTheme({
    boxShadows: {
        main: '0px 2px 4px rgba(129, 122, 254, 0.2)',
    },
    typography: {
        fontFamily: 'Sen, Roboto, sans-serif',
        fontWeightBold: '700',
        fontWeightMedium: '700',
        allVariants: { lineHeight: 1.4 },
        button: { lineHeight: 1.75 },
        h1: {
            fontSize: '1.5rem',
            lineHeight: 1.875,
        },
    },
    fontSizes: {
        mainHeader: '1.25rem',
        subHeader: '1.1rem',
        bodySize: '1rem',
        smallBody: '0.9rem',
        smallerBody: '0.8rem',
    },
    fontWeight: {
        thin: 300,
        medium: 400,
        semi: 700,
        bold: 700,
    },
    shape: {
        borderRadius: '3px',
        borderRadiusLarge: '10px',
        borderRadiusExtraLarge: '25px',
    },
    palette: {
        primary: {
            main: colors.purple[800],
            light: colors.purple[700],
            dark: colors.purple[900],
        },
        secondary: {
            main: colors.purple[800],
            light: colors.purple[700],
            dark: colors.purple[900],
        },
        info: {
            light: colors.blue[700],
            main: colors.blue[700],
            dark: colors.blue[800],
        },
        success: {
            light: colors.green[700],
            main: colors.green[700],
            dark: colors.green[800],
        },
        warning: {
            light: colors.orange[700],
            main: colors.orange[700],
            dark: colors.orange[800],
        },
        error: {
            light: colors.red[700],
            main: colors.red[700],
            dark: colors.red[800],
        },
        divider: colors.grey[300],
        dividerAlternative: colors.grey[500],
        grey: colors.grey,
        text: {
            primary: colors.grey[900],
            secondary: colors.grey[800],
            disabled: colors.grey[600],
        },
        code: {
            main: '#0b8c8f',
            diffAdd: 'green',
            diffSub: 'red',
            diffNeutral: 'black',
            edited: 'blue',
            background: '#efefef',
        },
        activityIndicators: {
            unknown: colors.grey[100],
            recent: colors.green[100],
            inactive: colors.orange[200],
            abandoned: colors.red[200],
        },
    },
    components: {
        MuiTableHead: {
            styleOverrides: {
                root: {
                    background: colors.grey[200],
                },
            },
        },
        MuiLink: {
            styleOverrides: {
                root: {
                    color: colors.purple[900],
                },
            },
        },
        MuiBreadcrumbs: {
            styleOverrides: {
                root: {
                    '& a': {
                        color: colors.purple[900],
                    },
                },
            },
        },
        MuiTableRow: {
            styleOverrides: {
                root: {
                    '&.MuiTableRow-hover:hover': {
                        background: colors.grey[100],
                    },
                },
            },
        },
    },
});
