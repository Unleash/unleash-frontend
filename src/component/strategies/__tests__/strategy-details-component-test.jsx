import React from 'react';
import { ThemeProvider } from '@material-ui/core';
import StrategyDetails from '../strategy-details-component';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import theme from '../../../themes/main-theme';
import { AccessProvider } from '../../Access/access-context';
import { createFakeStore } from '../../../accessStoreFake';

test('renders correctly with one strategy', () => {
    const strategy = {
        name: 'Another',
        description: "another's description",
        editable: true,
        parameters: [
            {
                name: 'customParam',
                type: 'list',
                description: 'customList',
                required: true,
            },
        ],
    };
    const applications = [
        {
            appName: 'appA',
            description: 'app description',
        },
    ];
    const toggles = [
        {
            name: 'toggleA',
            description: 'toggle description',
        },
    ];
    const tree = renderer.create(
        <MemoryRouter>
            <AccessProvider store={createFakeStore()}>
            <ThemeProvider theme={theme}>
                    <StrategyDetails
                        strategyName={'Another'}
                        strategy={strategy}
                        activeTab="view"
                        applications={applications}
                        toggles={toggles}
                        fetchStrategies={jest.fn()}
                        fetchApplications={jest.fn()}
                        fetchFeatureToggles={jest.fn()}
                        history={{}}
                    />
            </ThemeProvider>
            </AccessProvider>
        </MemoryRouter>
    );

    expect(tree).toMatchSnapshot();
});
