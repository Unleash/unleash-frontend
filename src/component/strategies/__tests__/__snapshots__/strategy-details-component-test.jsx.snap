// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with one strategy 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
>
  <div
    className="makeStyles-headerContainer-1"
  >
    <div
      className="makeStyles-headerTitleContainer-5"
    >
      <div
        className=""
      >
        <h2
          className="MuiTypography-root makeStyles-headerTitle-6 MuiTypography-h2"
        >
          Another
        </h2>
      </div>
    </div>
  </div>
  <div
    className="makeStyles-bodyContainer-2"
  >
    <div
      className="MuiGrid-root MuiGrid-container"
    >
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-12"
      >
        <h6
          className="MuiTypography-root MuiTypography-subtitle1"
        >
          another's description
        </h6>
        <div>
          <div
            className="MuiPaper-root makeStyles-tabNav-8 MuiPaper-elevation1 MuiPaper-rounded"
          >
            <div
              className="MuiTabs-root"
            >
              <div
                className="MuiTabs-scroller MuiTabs-fixed"
                onScroll={[Function]}
                style={
                  Object {
                    "marginBottom": null,
                    "overflow": "hidden",
                  }
                }
              >
                <div
                  className="MuiTabs-flexContainer MuiTabs-centered"
                  onKeyDown={[Function]}
                  role="tablist"
                >
                  <button
                    aria-controls="tabpanel-0"
                    aria-selected={true}
                    className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected"
                    disabled={false}
                    id="tab-0"
                    onBlur={[Function]}
                    onClick={[Function]}
                    onDragLeave={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    onMouseDown={[Function]}
                    onMouseLeave={[Function]}
                    onMouseUp={[Function]}
                    onTouchEnd={[Function]}
                    onTouchMove={[Function]}
                    onTouchStart={[Function]}
                    role="tab"
                    tabIndex={0}
                    type="button"
                  >
                    <span
                      className="MuiTab-wrapper"
                    >
                      Details
                    </span>
                    <span
                      className="MuiTouchRipple-root"
                    />
                  </button>
                  <button
                    aria-controls="tabpanel-1"
                    aria-selected={false}
                    className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary"
                    disabled={false}
                    id="tab-1"
                    onBlur={[Function]}
                    onClick={[Function]}
                    onDragLeave={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    onMouseDown={[Function]}
                    onMouseLeave={[Function]}
                    onMouseUp={[Function]}
                    onTouchEnd={[Function]}
                    onTouchMove={[Function]}
                    onTouchStart={[Function]}
                    role="tab"
                    tabIndex={-1}
                    type="button"
                  >
                    <span
                      className="MuiTab-wrapper"
                    >
                      Edit
                    </span>
                    <span
                      className="MuiTouchRipple-root"
                    />
                  </button>
                </div>
                <span
                  className="PrivateTabIndicator-root-9 PrivateTabIndicator-colorPrimary-10 MuiTabs-indicator"
                  style={
                    Object {
                      "left": 0,
                      "width": 0,
                    }
                  }
                />
              </div>
            </div>
          </div>
          <div>
            <div
              aria-labelledby="wrapped-tab-0"
              hidden={false}
              id="wrapped-tabpanel-0"
              role="tabpanel"
            >
              <div
                className="listcontainer"
              >
                <div
                  className="MuiGrid-root MuiGrid-container"
                >
                  <div
                    className="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-12 MuiGrid-grid-md-12"
                  >
                    <h6>
                      Parameters
                    </h6>
                    <hr />
                    <ul
                      className="MuiList-root MuiList-padding"
                    >
                      <li
                        className="MuiListItem-root MuiListItem-gutters"
                        disabled={false}
                      >
                        <div
                          aria-describedby={null}
                          className="MuiListItemAvatar-root"
                          onBlur={[Function]}
                          onFocus={[Function]}
                          onMouseLeave={[Function]}
                          onMouseOver={[Function]}
                          onTouchEnd={[Function]}
                          onTouchStart={[Function]}
                          title="Required"
                        >
                          <span
                            aria-hidden={true}
                            className="material-icons MuiIcon-root"
                          >
                            add
                          </span>
                        </div>
                        <div
                          className="MuiListItemText-root MuiListItemText-multiline"
                        >
                          <span
                            className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                          >
                            <div>
                              customParam
                               
                              <small>
                                (
                                list
                                )
                              </small>
                            </div>
                          </span>
                          <p
                            className="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
                          >
                            customList
                          </p>
                        </div>
                      </li>
                    </ul>
                  </div>
                  <div
                    className="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-12 MuiGrid-grid-md-6"
                  >
                    <h6>
                      Applications using this strategy
                    </h6>
                    <hr />
                    <ul
                      className="MuiList-root MuiList-padding"
                    >
                      <li
                        className="MuiListItem-root listItem MuiListItem-gutters"
                        disabled={false}
                      >
                        <div
                          className="MuiListItemAvatar-root"
                        >
                          <div
                            className="MuiAvatar-root MuiAvatar-circle MuiAvatar-colorDefault"
                          >
                            <span
                              aria-hidden={true}
                              className="material-icons MuiIcon-root"
                            >
                              apps
                            </span>
                          </div>
                        </div>
                        <div
                          className="MuiListItemText-root MuiListItemText-multiline"
                        >
                          <span
                            className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                          >
                            <a
                              className="listLink truncate"
                              href="/applications/appA"
                              onClick={[Function]}
                            >
                              appA
                            </a>
                          </span>
                          <p
                            className="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
                          >
                            app description
                          </p>
                        </div>
                      </li>
                    </ul>
                  </div>
                  <div
                    className="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-12 MuiGrid-grid-md-6"
                  >
                    <h6>
                      Toggles using this strategy
                    </h6>
                    <hr />
                    <ul
                      className="MuiList-root truncate MuiList-padding"
                      style={
                        Object {
                          "textAlign": "left",
                        }
                      }
                    >
                      <li
                        className="MuiListItem-root MuiListItem-gutters"
                        disabled={false}
                      >
                        <div
                          aria-describedby={null}
                          className="MuiListItemAvatar-root"
                          onBlur={[Function]}
                          onFocus={[Function]}
                          onMouseLeave={[Function]}
                          onMouseOver={[Function]}
                          onTouchEnd={[Function]}
                          onTouchStart={[Function]}
                          title="Disabled"
                        >
                          <span
                            aria-hidden={true}
                            className="material-icons MuiIcon-root"
                          >
                            pause
                          </span>
                        </div>
                        <div
                          className="MuiListItemText-root MuiListItemText-multiline"
                        >
                          <span
                            className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                          >
                            <a
                              href="/features/view/toggleA"
                              onClick={[Function]}
                            >
                              toggleA
                            </a>
                          </span>
                          <p
                            className="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
                          >
                            toggle description
                          </p>
                        </div>
                      </li>
                    </ul>
                  </div>
                </div>
              </div>
            </div>
            <div
              aria-labelledby="wrapped-tab-1"
              hidden={true}
              id="wrapped-tabpanel-1"
              role="tabpanel"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
